pyhf/pdf.py

```diff
@@ -126,7 +128,44 @@ def __init__(self, spec, **config_kwargs):
         # build up our representation of the specification
         self.config = _ModelConfig.from_spec(self.spec,**config_kwargs)
 
+    def _mtype_results(self,mtype,pars):
+        mtype_results = {}
+        for channel in self.spec['channels']:
+            for sample in channel['samples']:
+                for mname in sample['modifiers_by_type'].get(mtype,[]):
+                    modifier = self.config.modifier(mname)
+                    modpars  = pars[self.config.par_slice(mname)]
+                    mtype_results.setdefault(channel['name'],
+                            {}).setdefault(sample['name'],
+                            []).append(
+                                modifier.apply(channel, sample, modpars)
+                            )
+        return mtype_results
+
```

Comment:
This method implements the computation off the modifier application for a given modifier type and within that type for each channel and sample..

in a follow up PR it will be further refactored to reverse the order of the three loops, such that the outer-most loop is over modifiers (which is the structure we are aiming for in #251

Reply:
the refactored code will look like this

```
    def _mtype_results(self,mtype,pars):
        mtype_results = {}
        if mtype in self.combined_mods.keys():
            return self.combined_mods[mtype].apply(pars)
        for channel in self.spec['channels']:
            for sample in channel['samples']:
                for mname in sample['modifiers_by_type'].get(mtype,[]):
                    modifier = self.config.modifier(mname)
                    modpars  = pars[self.config.par_slice(mname)]
                    mtype_results.setdefault(channel['name'],
                            {}).setdefault(sample['name'],
                            []).append(
                                modifier.apply(channel, sample, modpars)
                            )
        return mtype_results
```

Reply:
Isn't this before refactoring? It's what you have right now in this commit.

Reply:
check the `if` clause which uses the combined mods which are coming.. it might be that e will have combined mods for each possible mtype so the else might never be reached

Reply:
I see.